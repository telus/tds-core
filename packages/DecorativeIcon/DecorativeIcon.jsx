import React from 'react'
import PropTypes from 'prop-types'

import Icon from '../../shared/components/Icon/Icon'

/**
 * An icon used for visual aesthetics only.
 */
const DecorativeIcon = ({ symbol, variant, size, ...rest }) => (
  <Icon {...rest} symbol={symbol} variant={variant} size={size} aria-hidden="true" />
)

DecorativeIcon.propTypes = {
  /**
   * Name of the icon symbol.
   */
  symbol: PropTypes.oneOf([
    'caretDown',
    'caretUp',
    'checkmark',
    'chevron',
    'leftChevron',
    'exclamationPointCircle',
    'expander',
    'hamburger',
    'location',
    'minus',
    'plus',
    'questionMarkCircle',
    'spyglass',
    'times',
  ]).isRequired,
  /**
   * The appearance of the Icon.
   */
  variant: PropTypes.oneOf(['primary', 'secondary', 'inverted', 'error']),
  /**
   * The icon size in pixels.
   */
  size: PropTypes.oneOf([16, 24, 48]),
}

DecorativeIcon.defaultProps = {
  variant: undefined,
  size: 24,
}

export default DecorativeIcon
