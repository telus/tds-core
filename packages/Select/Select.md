### Usage criteria

* Include a `placeholder` to provide instructions such as "Please selectâ€¦" as an unselectabale option (recommended)
* Use when options are between 7-15 (recommended)
* For options are 6 and under, use [`radio`](#radio) groups
* For options 15 and above, look for alternatives (e.g grouping into categories, or input field with auto-complete etc.)

[Reference](https://baymard.com/blog/drop-down-usability)

```
<Select
  label="Province"
  placeholder="Please select..."
  options={[{text: 'Alberta', value: 'AB'}, {text: 'British Columbia', value: 'BC'}, {text: 'Ontario', value: 'ON'}, {text: 'Quebec', value: 'QC'}]}
/>
```
