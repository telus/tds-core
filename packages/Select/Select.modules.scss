@import '../colour/colours';

.wrapper {
  composes: relative from '../../shared/styles/Position.modules.scss';

  background-color: $color-white;
}

.select {
  composes: height from '../../shared/styles/Forms.modules.scss';

  // Hide the arrows
  appearance: none;

  // Hide the arrows for IE
  &::-ms-expand {
    display: none;
  }

  background-color: transparent;
}

.positionSelectOnTop {
  composes: relative from '../../shared/styles/Position.modules.scss';

  z-index: 2;
}

.iconsPosition {
  composes: absolute centerVertically from '../../shared/styles/Position.modules.scss';
  composes: iconRight from '../../shared/components/FormField/FormField.modules.scss';

  z-index: 3;
  pointer-events: none;
}
