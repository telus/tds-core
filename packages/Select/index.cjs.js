const Select = require('./dist/index.cjs')

module.exports = Select
