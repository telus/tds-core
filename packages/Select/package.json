{
  "name": "@tds/core-select",
  "version": "1.0.13",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telusdigital/tds.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telusdigital/tds/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": ">=15",
    "react-dom": ">=15"
  },
  "dependencies": {
    "@tds/core-box": "^1.0.1",
    "@tds/core-decorative-icon": "^1.0.0",
    "@tds/core-input-feedback": "^1.0.2",
    "@tds/core-paragraph": "^1.0.4",
    "@tds/core-standalone-icon": "^1.0.0",
    "@tds/core-text": "^1.0.4",
    "@tds/util-prop-types": "^1.1.0",
    "prop-types": "^15.5.10"
  },
  "devDependencies": {
    "@tds/core-colours": "^1.0.0",
    "@tds/shared-animation": "^1.0.2",
    "@tds/shared-form-field": "^1.0.9",
    "react-transition-group": "^2.2.1"
  }
}
