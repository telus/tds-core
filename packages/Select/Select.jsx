import React, { useState } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Box from '@tds/core-box'
import DecorativeIcon from '@tds/core-decorative-icon'
import { FeedbackIcon } from '@tds/shared-form-field'
import Text from '@tds/core-text'
import Paragraph from '@tds/core-paragraph'
import { componentWithName } from '@tds/util-prop-types'
import InputFeedback from '@tds/core-input-feedback'

import { borders, position, forms } from '@tds/shared-styles'
import {
  colorShuttleGrey,
  colorWhite,
  colorPrimary,
  colorCardinal,
  colorAthensGrey,
} from '@tds/core-colours'
import { medium, mediumFont, color } from '@tds/shared-typography'

import safeRest from '../../shared/utils/safeRest'
import generateId from '../../shared/utils/generateId/generateId'

const SelectWrapper = styled.div({
  backgroundColor: colorWhite,
  ...position.relative,
})

const StyledSelect = styled.select(
  {
    ...forms.height,
    appearance: 'none',

    '&::-ms-expand': {
      display: 'none',
    },
    width: '100%',
    margin: 0,
    outline: 0,
    textOverflow: 'ellipsis',
    borderColor: colorShuttleGrey,
    backgroundColor: colorWhite,

    '&::placeholder': {
      font: 'inherit',
      letterSpacing: 'inherit',
      lineHeight: 'inherit',
      color: colorShuttleGrey,
    },
  },
  ({ showFeedbackIcon }) => ({
    padding: `0.5rem ${showFeedbackIcon ? '4' : '3'}rem 0.5rem 1rem`,
  }),
  borders.thin,
  borders.rounded,
  forms.font,
  medium,
  mediumFont,
  color,
  ({ withFeedbackIcon }) => ({
    '&::-webkit-inner-spin-button, &::-webkit-outer-spin-button': {
      appearance: 'none',
      margin: 0,
    },
    '-moz-appearance': 'textfield',
    minHeight: '3.25rem',
    maxHeight: '3.25rem',
    padding: withFeedbackIcon ? '0.5rem 3rem 0.5rem 1rem' : '0.5rem 1rem',
  }),
  {
    '&:focus': {
      borderColor: 'transparent',
      boxShadow: `0 0 4px 1px ${colorShuttleGrey}`,
      backgroundColor: colorWhite,
    },
  },
  ({ feedback }) => {
    let borderColor
    if (feedback === 'success') {
      borderColor = colorPrimary
    } else if (feedback === 'error') {
      borderColor = colorCardinal
    }
    return {
      '&:not(:focus)': {
        borderColor,
      },
    }
  },
  ({ disabled }) => {
    if (disabled) {
      return {
        backgroundColor: colorAthensGrey,
        borderColor: 'transparent',
        '&:not(:focus)': {
          borderColor: 'transparent',
        },
      }
    }
    return {}
  }
)

const IconWrapper = styled(Box)({
  ...position.absolute,
  top: '50%',
  transform: 'translateY(-50%)',
  right: '1rem',
  pointerEvents: 'none',
})

const IconLineFix = styled.div({
  lineHeight: 1,
})

const StyledLabelContainer = styled(Box)({
  alignItems: 'center',
})

const showFeedbackIcon = feedback => feedback === 'error' || feedback === 'success'
const renderHint = (hint, Component, id) => (
  <Component id={id} size="small">
    {hint}
  </Component>
)
const renderError = (error, errorId) => (
  <InputFeedback id={errorId} feedback="error">
    <Paragraph size="small">{error}</Paragraph>
  </InputFeedback>
)

const renderLabel = (id, label, hint, hintId, tooltip) => (
  <Box inline between="space-between">
    <label htmlFor={id || generateId(label).identity()}>
      <StyledLabelContainer inline tag="span" between={2}>
        <Text size="medium" bold data-testid="selectLabel">
          {label}
        </Text>
        {hint && renderHint(hint, Text, hintId)}
      </StyledLabelContainer>
    </label>
    {tooltip && React.cloneElement(tooltip, { connectedFieldLabel: label })}
  </Box>
)

const renderHelper = (helper, helperId, feedback, value) => {
  if (typeof helper === 'function') {
    return (
      <div id={helperId}>
        <Text size="small">{helper(feedback, value)}</Text>
      </div>
    )
  }

  return (
    <InputFeedback id={helperId} feedback={feedback}>
      <Text size="small">{helper}</Text>
    </InputFeedback>
  )
}

/**
 * @version ./package.json
 */
const Select = React.forwardRef(
  (
    {
      id,
      options,
      placeholder,
      label,
      hint,
      value,
      defaultValue,
      feedback,
      error,
      helper,
      tooltip,
      ...rest
    },
    ref
  ) => {
    const [isFocused, setIsFocused] = useState(false)
    const fieldId = generateId(id, rest.name, label)
    const errorId = error && feedback === 'error' && fieldId.postfix('error-message')
    const helperId = helper && fieldId.postfix('helper')
    const hintId = (hint && fieldId.postfix('hint')) || undefined

    const handleFocus = e => {
      setIsFocused(true)
      if (rest.onFocus) {
        rest.onFocus(e)
      }
    }

    const handleBlur = e => {
      setIsFocused(false)
      if (rest.onBlur) {
        rest.onBlur(e)
      }
    }

    return (
      <Box between={2}>
        {renderLabel(fieldId.identity(), label, hint, hintId, tooltip)}
        {helper && renderHelper(helper, helperId, feedback, value)}
        {error && feedback === 'error' && renderError(error, errorId)}
        <div css={{ position: 'relative' }}>
          <SelectWrapper>
            <StyledSelect
              ref={ref}
              id={fieldId.identity()}
              value={value}
              feedback={feedback}
              aria-invalid={feedback === 'error'}
              aria-describedby={errorId || hintId || helperId || undefined}
              onFocus={handleFocus}
              onBlur={handleBlur}
              defaultValue={value !== undefined ? undefined : defaultValue}
              {...safeRest(rest)}
            >
              {placeholder && (
                <option value="" disabled>
                  {placeholder}
                </option>
              )}
              {options.map(({ text, value: optValue }) => (
                <option key={optValue} value={optValue}>
                  {text}
                </option>
              ))}
            </StyledSelect>

            {!rest.disabled && (
              <IconWrapper inline between={3}>
                <FeedbackIcon
                  showIcon={showFeedbackIcon(feedback) && !isFocused}
                  feedback={feedback}
                />
                <IconLineFix>
                  <DecorativeIcon
                    symbol="caretDown"
                    variant={feedback === 'error' ? 'error' : 'primary'}
                    size={16}
                  />
                </IconLineFix>
              </IconWrapper>
            )}
          </SelectWrapper>
        </div>
      </Box>
    )
  }
)

Select.displayName = 'Select'

Select.propTypes = {
  /**
   * The id.
   */
  id: PropTypes.string,
  /**
   * The selectable options.
   */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string,
      value: PropTypes.string,
    })
  ).isRequired,
  /**
   * The label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Clarify the expected input.
   */
  hint: PropTypes.string,
  /**
   * Show a un-selectable initial option.
   */
  placeholder: PropTypes.string,
  /**
   * The value. Set this prop to change selected option.
   *
   * Only one of defaultValue or value should be set / mutated.
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * The default value. Set this prop to set the inital selected option.
   *
   * Only one of defaultValue or value should be set / mutated. `value` will override `defaultValue`.
   */
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * A feedback state.
   */
  feedback: PropTypes.oneOf(['success', 'error']),
  /**
   * An error message. Either an error or a helper should be used, not both.
   */
  error: PropTypes.string,
  /**
   * A detailed explanation of the input expected by a form field. Can be text,
   * other components, or HTML elements. Only appears when `feedback` is set to `error`.
   *
   * If a function is provided, it must return an `InputFeedback` component. The function will be
   * invoked with the following arguments.
   *
   * @param {String} feedback The input's current feedback state.
   * @param {String} value The input's current value.
   */
  helper: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  /**
   * A `Tooltip`
   */
  tooltip: componentWithName('Tooltip'),
}

Select.defaultProps = {
  id: undefined,
  hint: undefined,
  placeholder: undefined,
  value: undefined,
  defaultValue: '',
  feedback: undefined,
  error: undefined,
  helper: undefined,
  tooltip: undefined,
}

export default Select
