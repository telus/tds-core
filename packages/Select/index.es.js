import Select from './dist/index.es'

export default Select
