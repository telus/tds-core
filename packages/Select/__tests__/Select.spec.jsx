import React from 'react'
import { render, mount } from 'enzyme'

import StandaloneIcon from '@tds/core-standalone-icon'
import Text from '@tds/core-text'
import Paragraph from '@tds/core-paragraph'
import InputFeedback from '@tds/core-input-feedback'
import DecorativeIcon from '@tds/core-decorative-icon'
import { Fade } from '@tds/shared-animation'
import Tooltip from '../../Tooltip'

import Select from '../Select'

describe('Select', () => {
  const defaultProps = {
    label: 'The select',
    options: [
      { text: 'text one', value: 'value-1', readOnly: true },
      { text: 'text two', value: 'value-2', readOnly: true },
    ],
  }

  const doRender = (overrides = {}) => render(<Select {...defaultProps} {...overrides} />)

  const doMount = (overrides = {}) => {
    const select = mount(<Select {...defaultProps} {...overrides} />)

    const findSelectElement = () => select.find('select')

    return {
      select,
      label: select.find('label'),
      findSelectElement,
      findFeedbackIconFade: () => select.find(Fade),
      findHelper: () => select.find(InputFeedback),
      focus: (focusEvent = {}) => findSelectElement().simulate('focus', focusEvent),
      blur: (blurEvent = {}) => findSelectElement().simulate('blur', blurEvent),
    }
  }

  it('renders', () => {
    const select = doRender()

    expect(select).toMatchSnapshot()
  })

  it('renders with a feedback state and icon', () => {
    const select = doRender({ feedback: 'error' })

    expect(select).toMatchSnapshot()
  })

  it('needs a set of options', () => {
    const { findSelectElement } = doMount({
      options: [
        { text: 'some text', value: 'value-1', readOnly: true },
        { text: 'other text', value: 'value-2', readOnly: true },
      ],
    })

    expect(findSelectElement).toMatchSnapshot()
  })

  it('can render groups', () => {
    const { findSelectElement } = doMount({
      options: [
        { text: 'some text', value: 'value-1', readOnly: true },
        { text: 'other text', value: 'value-2', readOnly: true },
        {
          text: 'a group',
          options: [
            { text: 'a group option', value: 'group-value-1' },
            { text: 'other group option', value: 'group-value-2' },
          ],
        },
      ],
    })

    expect(findSelectElement).toMatchSnapshot()
  })

  it('positions the down caret so that the text does not overlap it', () => {
    const { findSelectElement } = doMount()

    expect(findSelectElement).toMatchSnapshot()
  })

  describe('label', () => {
    it('must have a label', () => {
      const { label } = doMount({ label: 'The label' })

      expect(label).toMatchSnapshot()
    })

    it('can have a short hint', () => {
      const { label } = doMount({ hint: 'The short hint' })

      expect(label).toMatchSnapshot()
    })

    it('can have a hint below', () => {
      const { label } = doMount({ hint: 'The short hint', hintPosition: 'below' })

      expect(label).toMatchSnapshot()
    })
  })

  describe('connecting the label to the select', () => {
    it('connects the label to the select', () => {
      const { label, findSelectElement } = doMount()

      expect(label.prop('htmlFor')).toEqual(findSelectElement().prop('id'))
    })

    it('uses the id when provided', () => {
      const { label, findSelectElement } = doMount({
        id: 'the-id',
        name: 'the-name',
        label: 'The label',
      })

      expect(label).toHaveProp('htmlFor', 'the-id')
      expect(findSelectElement()).toHaveProp('id', 'the-id')
    })

    it('uses the name when no id is provided', () => {
      const { label, findSelectElement } = doMount({ name: 'the-name', label: 'The label' })

      expect(label).toHaveProp('htmlFor', 'the-name')
      expect(findSelectElement()).toHaveProp('id', 'the-name')
    })

    it('generates an id from the label when no id or name is provided', () => {
      const { label, findSelectElement } = doMount({ label: 'The label' })

      expect(label).toHaveProp('htmlFor', 'the-label')
      expect(findSelectElement()).toHaveProp('id', 'the-label')
    })
  })

  describe('editability', () => {
    it('supports string values or number values', () => {
      let findSelectElement = doMount().findSelectElement
      expect(findSelectElement()).toHaveValue(undefined)

      findSelectElement = doMount({ value: 'some value', readOnly: true }).findSelectElement
      expect(findSelectElement()).toHaveValue('some value')
    })

    it('can receive a new value from a parent component', () => {
      const { select, findSelectElement } = doMount({ value: 'initial value', readOnly: true })

      select.setProps({ value: 'new value' })

      expect(findSelectElement()).toHaveValue('new value')
    })
  })

  describe('feedback states', () => {
    it('renders a checkmark icon on success feedback state', () => {
      const { findFeedbackIconFade } = doMount({ feedback: 'success' })

      expect(findFeedbackIconFade()).toContainReact(
        <StandaloneIcon
          symbol="checkmark"
          variant="primary"
          size={16}
          a11yText="The value of this input field is valid."
        />
      )
    })

    it('renders an error icon on error feedback state', () => {
      const { select, findFeedbackIconFade } = doMount({ feedback: 'error' })

      expect(select.find(DecorativeIcon)).toHaveProp('variant', 'error')
      expect(findFeedbackIconFade()).toContainReact(
        <StandaloneIcon
          symbol="exclamationPointCircle"
          variant="error"
          size={16}
          a11yText="The value of this input field is invalid."
        />
      )
    })

    it('fades the feedback icon in on focus lost and out on focus gained', () => {
      const { findFeedbackIconFade, focus } = doMount({ feedback: 'success' })

      expect(findFeedbackIconFade()).toHaveProp('in', true)

      focus()
      expect(findFeedbackIconFade()).toHaveProp('in', false)
    })

    it('ensures that the contents do not overlap the icons', () => {
      const { findSelectElement } = doMount({ feedback: 'success' })

      expect(findSelectElement).toMatchSnapshot()
    })
  })

  describe('disabling', () => {
    it('deactivates the select', () => {
      let findSelectElement = doMount().findSelectElement
      expect(findSelectElement).toMatchSnapshot()
      expect(findSelectElement()).not.toBeDisabled()

      findSelectElement = doMount({ disabled: true }).findSelectElement
      expect(findSelectElement()).toHaveProp('disabled')
      expect(findSelectElement()).toBeDisabled()
    })

    it('removes the positioning of the select so that the disabled treatment is displayed', () => {
      const { findSelectElement } = doMount({ disabled: true })

      expect(findSelectElement).toMatchSnapshot()
    })

    it('hides any icons', () => {
      const { select } = doMount({ disabled: true, feedback: 'error' })

      expect(select.find(StandaloneIcon)).not.toExist()
      expect(select.find(DecorativeIcon)).not.toExist()
    })
  })

  it('can see placeholder', () => {
    const { findSelectElement } = doMount({ placeholder: 'Please select...' })

    expect(findSelectElement()).toContainReact(
      <option value="" disabled>
        Please select...
      </option>
    )
  })

  it('can have an error message', () => {
    const { select } = doMount({
      id: 'some-id',
      error: 'Oh no a terrible error!',
      feedback: 'error',
    })

    expect(select).toMatchSnapshot()
  })

  describe('helpers', () => {
    it('can have a simple helper of some components', () => {
      const helper = <div>Some helper text.</div>
      const { select } = doMount({ id: 'some-id', helper })

      expect(select).toContainReact(
        <InputFeedback id="some-id_helper">
          <Text size="small">
            <div>Some helper text.</div>
          </Text>
        </InputFeedback>
      )
    })

    it('styles itself based on the select feedback state', () => {
      const helper = <Paragraph>Some helper text.</Paragraph>

      const { findHelper } = doMount({ feedback: 'success', helper })

      expect(findHelper()).toHaveProp('feedback', 'success')
    })

    it('can have a complex helper function to give control to the consumer', () => {
      const helper = jest.fn()
      helper.mockReturnValue(<InputFeedback>Some helper text.</InputFeedback>)

      const { select } = doMount({
        id: 'some-id',
        value: 'current value',
        feedback: 'error',
        helper,
        readOnly: true,
      })

      expect(helper).toHaveBeenCalledWith('error', 'current value')
      expect(select).toContainReact(
        <div id="some-id_helper">
          <Text size="small">
            <InputFeedback>Some helper text.</InputFeedback>
          </Text>
        </div>
      )
    })
  })

  describe('accessibility', () => {
    it('marks the select as invalid when in the error feedback state', () => {
      let findSelectElement = doMount().findSelectElement
      expect(findSelectElement()).toHaveProp('aria-invalid', false)

      findSelectElement = doMount({ feedback: 'error' }).findSelectElement
      expect(findSelectElement()).toHaveProp('aria-invalid', true)
    })

    it('does not attach aria-describedby to the select field when no error or helper is present', () => {
      const { findSelectElement } = doMount({ error: undefined, helper: undefined })

      expect(findSelectElement()).toHaveProp('aria-describedby', undefined)
    })

    it('connects the error message to the select field for screen readers', () => {
      const { findSelectElement, findHelper } = doMount({
        id: 'some-field-id',
        error: 'An error message',
        feedback: 'error',
      })

      expect(findSelectElement()).toHaveProp('aria-describedby', 'some-field-id_error-message')
      expect(findHelper()).toHaveProp('id', 'some-field-id_error-message')
    })

    it('connects a simple helper to the select field for screen readers', () => {
      const helper = <Paragraph>Some helper text.</Paragraph>
      const { findSelectElement, findHelper } = doMount({
        id: 'some-field-id',
        helper,
      })
      expect(findSelectElement()).toHaveProp('aria-describedby', 'some-field-id_helper')
      expect(findHelper()).toHaveProp('id', 'some-field-id_helper')
    })

    it('connects a complex helper to the select field for screen readers', () => {
      const helper = () => <InputFeedback>Complex helper</InputFeedback>
      const { select, findSelectElement } = doMount({ id: 'some-field-id', helper })

      expect(findSelectElement()).toHaveProp('aria-describedby', 'some-field-id_helper')
      expect(select).toContainReact(
        <div id="some-field-id_helper">
          <Text size="small">
            <InputFeedback>Complex helper</InputFeedback>
          </Text>
        </div>
      )
    })
  })

  describe('tooltip prop', () => {
    it('connects to Select', () => {
      const select = mount(
        <Select
          label="Some field"
          options={[
            {
              text: 'hi',
              value: 'hi',
              readOnly: true,
            },
          ]}
          tooltip={<Tooltip copy="en">The tooltip</Tooltip>}
        />
      )

      expect(select).toContainReact(
        <Tooltip copy="en" connectedFieldLabel="Some field">
          The tooltip
        </Tooltip>
      )
    })
  })

  it('passes additional attributes to the select element', () => {
    const { findSelectElement } = doMount({ name: 'a name', id: 'the-id' })

    expect(findSelectElement()).toHaveProp('name', 'a name')
    expect(findSelectElement()).toHaveProp('id', 'the-id')
  })

  it('does not allow custom CSS', () => {
    const { findSelectElement } = doMount({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(findSelectElement()).not.toHaveClassName('my-custom-class')
    expect(findSelectElement()).not.toHaveStyle('color', 'hotpink')
  })
})
