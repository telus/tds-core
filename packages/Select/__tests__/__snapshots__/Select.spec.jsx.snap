// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders 1`] = `
<div
  class="TDS_Box-modules__betweenBottomMargin-2___31zX_ TDS_Box-modules__stack___33m4D"
>
  <div
    class="row justifyContentSpaceBetween relative"
  >
    <label
      for="the-select"
    >
      <span
        class="TDS_Box-modules__betweenRightMargin-2___20pzC TDS_Box-modules__inline___jTHcz alignCenter"
      >
        <span
          class="TDS_Typography-modules__medium___1rxfE TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__boldFont___2YS3O TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__color___2CNH8"
        >
          The select
        </span>
      </span>
    </label>
  </div>
  <div
    class="wrapper"
  >
    <select
      aria-invalid="false"
      class="default select"
      id="the-select"
      style="padding-right:calc(16px + 2rem)"
    >
      <option
        value="value-1"
      >
        text one
      </option>
      <option
        value="value-2"
      >
        text two
      </option>
    </select>
    <div
      class="TDS_Box-modules__betweenRightMargin-3___1dOvx TDS_Box-modules__inline___jTHcz iconsPosition"
    >
      <div
        class="fixLineHeight"
      >
        <i
          aria-hidden="true"
          class="TDS_Icon-modules__iconCaretDown___3by8b TDS_Icon-modules__icon___13xYd TDS_Icon-modules__primary___2UY-N TDS_Icon-modules__size16___1pm12"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Select renders with a feedback state and icon 1`] = `
<div
  class="TDS_Box-modules__betweenBottomMargin-2___31zX_ TDS_Box-modules__stack___33m4D"
>
  <div
    class="row justifyContentSpaceBetween relative"
  >
    <label
      for="the-select"
    >
      <span
        class="TDS_Box-modules__betweenRightMargin-2___20pzC TDS_Box-modules__inline___jTHcz alignCenter"
      >
        <span
          class="TDS_Typography-modules__medium___1rxfE TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__boldFont___2YS3O TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__color___2CNH8"
        >
          The select
        </span>
      </span>
    </label>
  </div>
  <div
    class="wrapper"
  >
    <select
      aria-invalid="true"
      class="error select"
      id="the-select"
      style="padding-right:calc(32px + 3rem)"
    >
      <option
        value="value-1"
      >
        text one
      </option>
      <option
        value="value-2"
      >
        text two
      </option>
    </select>
    <div
      class="TDS_Box-modules__betweenRightMargin-3___1dOvx TDS_Box-modules__inline___jTHcz iconsPosition"
    >
      <div
        style="transition:opacity 100ms ease-in-out;opacity:1"
      >
        <div
          class="fixLineHeight"
        >
          <i
            aria-label="The value of this input field is invalid."
            class="TDS_Icon-modules__iconExclamationPointCircle___1HgRN TDS_Icon-modules__icon___13xYd TDS_Icon-modules__error___yuB9x TDS_Icon-modules__size16___1pm12"
          />
        </div>
      </div>
      <div
        class="fixLineHeight"
      >
        <i
          aria-hidden="true"
          class="TDS_Icon-modules__iconCaretDown___3by8b TDS_Icon-modules__icon___13xYd TDS_Icon-modules__error___yuB9x TDS_Icon-modules__size16___1pm12"
        />
      </div>
    </div>
  </div>
</div>
`;
